File: gosdk-develop-backup-20260812/gosdk-develop/crypto/address.go
package crypto

import (
	"encoding/hex"
	"encoding/json"
	"fmt"
	"github.com/milon-labs/milon-go-sdk/postcard"
	"strings"

	"github.com/btcsuite/btcutil/base58"
)

// AddressRawLen raw byte length of a address
const AddressRawLen = 20

// AddressHexLen hexadecimal literal length of a address
const AddressHexLen = AddressRawLen * 2

// NewAddressFromPublicKey derives a address from a public key
func NewAddressFromPublicKey(pk *PublicKey) (*Address, error) {
	// 计算公钥的 BLAKE3 哈希（32字节）
	hash := Hash32([]byte(PkAddressDomainContext), pk.Bytes)

	// 取前20字节作为地址
	var digest [AddressRawLen]byte
	copy(digest[:], hash[:AddressRawLen])

	return &Address{Bytes: digest}, nil
}

// NewAddressFromBytes parses an address from a 20-byte slice
func NewAddressFromBytes(bytes []byte) (*Address, error) {
	if len(bytes) != AddressRawLen {
		return nil, fmt.Errorf("invalid address length: expected %d, got %d", AddressRawLen, len(bytes))
	}

	var digest [AddressRawLen]byte
	copy(digest[:], bytes)
	return &Address{Bytes: digest}, nil
}

// NewAddressFromStringRelaxed parses an address from a hex or base58 string
func NewAddressFromStringRelaxed(s string) (*Address, error) {
	s = strings.TrimSpace(s)

	hexBody := s
	if len(s) >= 2 && s[0] == '0' && (s[1] == 'x' || s[1] == 'X') {
		hexBody = s[2:]
	}
	if len(hexBody) == AddressHexLen {
		return newAddressFromHex(s)
	}

	return newAddressFromBase58(s)
}

func newAddressFromHex(s string) (*Address, error) {
	s = strings.TrimSpace(s)

	hexStr := s
	if len(s) >= 2 && s[0] == '0' && (s[1] == 'x' || s[1] == 'X') {
		hexStr = s[2:]
	}

	if len(hexStr) != AddressHexLen {
		return nil, fmt.Errorf("invalid hex length: expected %d, got %d", AddressHexLen, len(hexStr))
	}

	buf, err := hex.DecodeString(hexStr)
	if err != nil {
		return nil, fmt.Errorf("invalid hex string: %w", err)
	}

	return NewAddressFromBytes(buf)
}

func newAddressFromBase58(s string) (*Address, error) {
	buf := base58.Decode(s)

	if len(buf) != AddressRawLen {
		return nil, fmt.Errorf("invalid base58 decoded length: expected %d, got %d", AddressRawLen, len(buf))
	}

	return NewAddressFromBytes(buf)
}

type Address struct {
	Bytes [AddressRawLen]byte
}

func (a *Address) AsBytes() [AddressRawLen]byte {
	return a.Bytes
}

func (a *Address) ToHex() string {
	return hex.EncodeToString(a.Bytes[:])
}

func (a *Address) ToBase58() string {
	return base58.Encode(a.Bytes[:])
}

func (a *Address) MarshalJSON() ([]byte, error) {
	return json.Marshal(a.ToBase58())
}

func (a *Address) UnmarshalJSON(data []byte) error {
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}

	addr, err := NewAddressFromStringRelaxed(s)
	if err != nil {
		return fmt.Errorf("failed to parse address: %w", err)
	}

	*a = *addr
	return nil
}

// String implements the Stringer interface, returns base58 format
func (a Address) String() string {
	return a.ToBase58()
}

func (a *Address) MarshalPostcard(serializer *postcard.Serializer) error {
	serializer.SerializeFixedBytes(a.Bytes[:])
	return nil
}

func (a *Address) UnmarshalPostcard(deserializer *postcard.Deserializer) error {
	buf, err := deserializer.DeserializeFixedBytes(AddressRawLen)
	if err != nil {
		return fmt.Errorf("failed to deserialize address bytes: %w", err)
	}

	addr, err := NewAddressFromBytes(buf)
	if err != nil {
		return fmt.Errorf("failed to create address from bytes: %w", err)
	}

	*a = *addr
	return nil
}

File: gosdk-develop/python/milon_sdk/crypto/address.py
"""地址类型与派生（对应 Go crypto/address.go）。

- Address = BLAKE3(MILON_ROOT || "milon.address.pk.v1" || pk.Bytes)[:20]
- 序列化形态：Hex（40 字符）或 Base58（Go String() 默认 Base58）。
"""
from __future__ import annotations

import base58  # type: ignore

from ..postcard import Deserializer, Serializer
from .errors import InvalidPublicKeyError
from .hashes import PK_ADDRESS_DOMAIN, hash32

ADDRESS_RAW_LEN = 20
ADDRESS_HEX_LEN = ADDRESS_RAW_LEN * 2


def _decode_hex(s: str) -> bytes:
    body = s[2:] if len(s) >= 2 and s[:2] in ("0x", "0X") else s
    if len(body) != ADDRESS_HEX_LEN:
        raise ValueError(f"invalid hex length: expected {ADDRESS_HEX_LEN}, got {len(body)}")
    return bytes.fromhex(body)


class Address:
    """20 字节链上地址。以不可变 bytes 承载，可哈希（Go 中用作 map key）。"""

    __slots__ = ("_bytes",)

    def __init__(self, raw: bytes):
        if len(raw) != ADDRESS_RAW_LEN:
            raise ValueError(f"invalid address length: expected {ADDRESS_RAW_LEN}, got {len(raw)}")
        self._bytes = bytes(raw)

    # ---------------------------------------------------------- 构造
    @classmethod
    def from_public_key(cls, pk: "PublicKey") -> "Address":  # noqa: F821
        digest = hash32(PK_ADDRESS_DOMAIN, pk.as_bytes())
        return cls(digest[:ADDRESS_RAW_LEN])

    @classmethod
    def from_bytes(cls, raw: bytes) -> "Address":
        return cls(raw)

    @classmethod
    def from_hex(cls, s: str) -> "Address":
        return cls(_decode_hex(s.strip()))

    @classmethod
    def from_base58(cls, s: str) -> "Address":
        buf = base58.b58decode(s.strip())
        if len(buf) != ADDRESS_RAW_LEN:
            raise ValueError(
                f"invalid base58 decoded length: expected {ADDRESS_RAW_LEN}, got {len(buf)}"
            )
        return cls(buf)

    @classmethod
    def from_relaxed(cls, value: object) -> "Address":
        """接受 Address、bytes、hex 字符串（可带 0x）、base58 字符串。"""
        if isinstance(value, Address):
            return value
        if isinstance(value, bytes):
            return cls(value)
        if isinstance(value, str):
            s = value.strip()
            body = s[2:] if len(s) >= 2 and s[:2] in ("0x", "0X") else s
            if len(body) == ADDRESS_HEX_LEN:
                return cls.from_hex(body)
            return cls.from_base58(s)
        raise TypeError(f"unsupported address input type {type(value).__name__}")

    # ---------------------------------------------------------- 输出
    def as_bytes(self) -> bytes:
        return self._bytes

    def to_hex(self) -> str:
        return self._bytes.hex()

    def to_base58(self) -> str:
        return base58.b58encode(self._bytes).decode("ascii")

    def __str__(self) -> str:
        return self.to_base58()

    def __repr__(self) -> str:
        return f"Address({self.to_base58()})"

    def __bytes__(self) -> bytes:
        return self._bytes

    def __eq__(self, other: object) -> bool:
        if isinstance(other, Address):
            return self._bytes == other._bytes
        if isinstance(other, bytes):
            return self._bytes == other
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self._bytes)

    # ---------------------------------------------------------- 编解码
    def marshal_postcard(self, serializer: Serializer) -> None:
        serializer.write(self._bytes)

    @classmethod
    def unmarshal_postcard(cls, d: Deserializer) -> "Address":
        buf = d.deserialize_fixed_bytes(ADDRESS_RAW_LEN)
        return cls(buf)
